### DISCUSSION TITLE
数据并行数 ×  epoch数 =  真实的epoch数?

### DISCUSSION THREAD

**Discussion by bobo0810** [hpcaitech/ColossalAI#2961] [Community | Q&A]
每个gpu上的dataloader都是完整的数据集，未做拆分。 即epoch=3  gpu=2时仅数据并行，模型实际上过了6遍数据集。
```
engine, train_dataloader, val_dataloader, _ = colossalai.initialize(
        model,
        optimizer,
        criterion,
        train_dataloader,
        val_dataloader,
    )
```

**Comment by binmakeswell** [comment]
Hi @bobo0810 如果是PyTorch正常dataloader提供给Colossal，会被自动转成DistributedSampler。每个GPU各自处理一部分数据，共同完成整个数据集的epoch。epoch=3是3遍数据集。

**Reply by bobo0810** [reply]
> Hi @bobo0810 如果是PyTorch正常dataloader提供给Colossal，会被自动转成DistributedSampler。每个GPU各自处理一部分数据，共同完成整个数据集的epoch。epoch=3是3遍数据集。

Hi @binmakeswell 
数据并行=2  epoch=3,理论上每张图片只出现3次，但从日志上看 每轮epoch出现2次，共出现6次。

启动命令`colossalai run --nproc_per_node 2 train.py --epochs 3`

代码
![image](https://user-images.githubusercontent.com/26057879/222613864-cf2bd32c-6d09-4700-9b8b-0c97bf6c198b.png)

日志
![image](https://user-images.githubusercontent.com/26057879/222613711-9dbca871-a9f9-4d12-8940-140b22815101.png)
